Feature: Suggest implicit or explicit typing code styles.

balajikris opened the pull request:
#### Use var code style: ####

Implements #7062.

Implement analyzer and codefixers for suggesting implicit typing and for replacing usages of implicit typing with explicit typing based on user options. Also adds options information to capture user styles for using or not using implicit typing. 

this implementation adds the following _options_ to 'Code Style' options page: 
1. suggest or do not suggest var for primitive types
2. suggest or do not suggest var where typing information is evident in the declaration.
3. suggest or do not suggest var in other places not specified above.

###### screenshot: ######
![image](https://cloud.githubusercontent.com/assets/4213867/13201446/b80c2d46-d826-11e5-9749-0e18562e9ed8.png)


###### Tasks List: ######
- [x] Implement analyzers and fixers.
- [x] unit tests.
- [x] option pages UI.
- [x] Enforcement levels (right now, un-checking use var implies suggest explicit typing)
- [ ] ~~Pass options through to simplifier and code generation engine as appropriate~~
- [ ] ~~Finalize and freeze UI and strings~~
- [ ] ~~Implement Fix All~~
- [ ] ~~Ensure UI is friendly with screen readers.~~

###### Future Improvements: ######
- [ ] ~~Better options preview - update preview to reflect current set of feature related option settings instead of the currently selected option.~~
- [ ] ~~Go to option settings from light bulb preview~~

balajikris commented:
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/2933/
// Retest reason: seems like a machine issue, it failed with a build error whereas 4 other queues were able to build the change.

balajikris commented:
current unit test failure is unrelated to this as well, filed as issue #8060

balajikris commented:
@dotnet/roslyn-ide : Please take a look. Some tasks are pending, but I'd like to merge something and keep working on it incrementally. If you believe some part of the pending task needs to be done before checking in (say enforcement levels, as they govern how intrusive this feature can or cannot be :wink: ), please feel free to call it out.

balajikris commented:
I recommend viewing the files changed directly over the commit-by-commit approach for this. I have some commits that cancel out each other but I didn't remove 'em. Furthermore, all major changes are in the newly added files.

brettfo commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+{
+    internal abstract class CSharpTypingStyleDiagnosticAnalyzerBase : DiagnosticAnalyzer, IBuiltInAnalyzer
+    {
+        protected enum TypingStyles
```
`[Flags]`

paul1956 commented:
You are going to run into problems because ToDisplayString returns the wrong value in too many cases to list.

DustinCampbell commented:
>You are going to run into problems because ToDisplayString returns the wrong value in too many cases to list.

@paul1956 -- do you have some examples that you could list? The defaults for `ToDisplayString()` generally work pretty well for this sort of thing (with some exceptions) and we do it in several places in IDE code.

DustinCampbell commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,97 @@
+
+            var typeSymbol = semanticModel.GetTypeInfo(typeSyntax).ConvertedType;
+
+            var typeName = SyntaxFactory.ParseTypeName(typeSymbol.ToDisplayString())
```
What if the result of typeSymbol.ToDisplayString(...) isn't parseable (e.g. an anonymous type)? In those cases, `GetSpeculativeSymbolInfo(...)` can return something pretty wacky. A decent heuristic is to verify that `typeName` doesn't contain any skipped token trivia.

balajikris replied:
the fixer shouldn't be running on `anonymous types` as you could only use var to declare them, but sure, i see your point - I'll put in the heuristic, thanks!

rchande replied:
Does this work with generics?

DustinCampbell replied:
Sorry @balajikris  -- I missed that this was the fixer.

DustinCampbell replied:
Is there a place where anonymous types are filtered out by the diagnostic? I couldn't find it. Note: Anonymous types can appear in nefarious places, such as being passed through as generic type parameters. We have an extension method that checks for that [here] (http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/ITypeSymbolExtensions.cs,b2cea94feb884a75).

balajikris replied:
[here](https://github.com/dotnet/roslyn/pull/8045/files#diff-171791ab0cb57419599831446b6c9325R76) and I seem to be calling the same extension method that you pointed out?

balajikris replied:
@rchande : yes it does, I likely have a test for it too. If not, i'll add.

balajikris replied:
@DustinCampbell : Looks like `ParseTypeName` consumes unexpected tokens in the given string and sets `diagnostics`. I'm thinking of asserting that the parsed name does not contain diagnostics.  there is also a `TypeSyntaxExtensions.IsPotentialTypeName` which does some syntactic checks and some semantic checks as well. However, it feels a bit too much to me, since `ParseTypeName` seems to attach diagnostics to the node and it seems sufficient. Does that sound good?

rchande replied:
Could you use the code generation service to produce a `NameSyntax`?

CyrusNajmabadi replied:
Yes.  Do not use Parse here.  We already have *good* functions that do the right thing to convert a type to a piece of syntax.  Use those.

CyrusNajmabadi replied:
You're still calling Parse here.  Don't. :)

CyrusNajmabadi replied:
As a bonus, you won't need to add the simplifier annotation.

balajikris replied:
I totally missed this comment. due to the how long this pr has gotten. I'll take care of this.

DustinCampbell commented:
>1. suggest or do not suggest var for primitive types
2. suggest or do not suggest var where typing information is evident in the declaration.
3. suggest or do not suggest var in other places not specified above.

Could we chat offline about the options? The wording doesn't seem right to me.

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+            VarWhereApparent = 1 << 2,
+            NoVarWhereApparent = 1 << 3,
+            VarWherePossible = 1 << 4,
+            NoVarWherePossible = 1 << 5
```
This enum seems a little weird. Why do you need bits for when an option is set and for when its unset? Shouldn't the absence of the bit represent the option not being set?

balajikris replied:
awk! will remove :smile:

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+            return stylePreferences;
+        }
+
+        protected bool IsTypeApparentInDeclaration(SyntaxNode declarationStatement, SemanticModel semanticModel, TypingStyles stylePreferences, CancellationToken cancellationToken)
```
Could you add a comment that defines when a type is "apparent"? I'm a bit confused by what that means -- especially because of the foreach rule below.

balajikris replied:
hmm. good point. Seemingly I've mixed, two things - "where type information can be evident" and "where implicit typing is generally preferred, by the framework design guidelines and by developers"

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+                ? ((CheckedExpressionSyntax)initializer.Value).Expression
+                : initializer.Value;
+
+        private bool PossibleConversionMethod(IMethodSymbol methodSymbol, ITypeSymbol declaredType, SemanticModel semanticModel, ExpressionSyntax typeName, CancellationToken cancellationToken)
```
`IsPossibleConversionMethod(...)`

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+
+            // case: int.Parse or TextSpan.FromBounds
+            if (methodSymbol.Name.StartsWith("Parse", StringComparison.Ordinal) 
+                || methodSymbol.Name.StartsWith("From", StringComparison.Ordinal))
```
These are probably too general. I can easily imagine false positives (especially for "From"). If there are only a handful of methods we care about in the BCL and in Roslyn, why not just specifically look for those?

DustinCampbell replied:
Ah... maybe not. I missed what you're doing below. That should be good.

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+                    : declaredType.Name;
+
+            // case: Convert.ToString or iConvertible.ToChar
+            if (methodSymbol.Name.Equals("To" + declaredTypeName, StringComparison.Ordinal))
```
Again, if we only care about methods on `Convert` or `IConvertible`, maybe search for those specifically?

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+            }
+
+            var possibleMethodSymbol = semanticModel.GetSymbolInfo(memberName, cancellationToken).Symbol;
+            if (possibleMethodSymbol == null || !possibleMethodSymbol.IsKind(SymbolKind.Method))
```
You could probably clean this up a bit and combine the next variable declaration by doing something like...

```C#
var methodSymbol = semanticModel.GetSymbolInfo(memberName, cancellationToken).Symbol as IMethodSymbol;
if (methodSymbol == null)
{
    return false;
}
```

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+
+            if (declarationStatement.IsKind(SyntaxKind.VariableDeclaration))
+            {
+                var declaration = ((VariableDeclarationSyntax)declarationStatement);
```
unnecessary parentheses.

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+            }
+            else if (declarationStatement.IsKind(SyntaxKind.ForEachStatement))
+            {
+                var declaration = ((ForEachStatementSyntax)declarationStatement);
```
ditto

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,272 @@
+        {
+            None = 0,
+            VarForIntrinsic = 1 << 0,
+            NoVarForIntrinsic = 1 << 1,
```
Note: nowhere in the C# spec does the word "intrinsic" appear.

DustinCampbell commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,97 @@
+                  node.IsKind(SyntaxKind.IdentifierName) ||
+                  node.IsKind(SyntaxKind.GenericName) ||
+                  node.IsKind(SyntaxKind.AliasQualifiedName)))
+            {
```
```C#
if (node?.MatchesKind(SyntaxKind.PredefinedType, SyntaxKind.ArrayType, SyntaxKind.IdentifierName, SyntaxKind.GenericName, SyntaxKind.AliasQualifiedName) == false)
{
    return false;
}
```

Note that this would create a params array, so you might want to lift that out.

DustinCampbell commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,97 @@
+                return;
+            }
+
+            var node = token.GetAncestors<SyntaxNode>().First(n => n.Span.Contains(span));
```
Why not just ```root.FindNode(span)``` rather than bothering with the token?

CyrusNajmabadi replied:
+1

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseExplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,91 @@
+                || stylePreferences.HasFlag(TypingStyles.NoVarWherePossible) && !(isIntrinsicType || isTypeApparent);
+        }
+
+        protected override bool AnalyzeVariableDeclaration(TypeSyntax typeName, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken, out TextSpan issueSpan)
```
`TryAnalyzeVariableDeclaration(...)`

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+
+        public CSharpUseImplicitTypingDiagnosticAnalyzer() : base(s_descriptorUseImplicitTyping)
+        {
+
```
extra line

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpUseImplicitTypingDiagnosticAnalyzer : CSharpTypingStyleDiagnosticAnalyzerBase
+    {
+
```
extra line

CyrusNajmabadi replied:
exta line is still htere :)

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+
+            // If there exists a type named var, return.
+            var conflict = semanticModel.GetSpeculativeSymbolInfo(typeName.SpanStart, candidateReplacementNode, SpeculativeBindingOption.BindAsTypeOrNamespace).Symbol;
+            if (conflict != null && conflict.IsKind(SymbolKind.NamedType))
```
```C#
if (conflict?.IsKind(SymbolKind.NamedType) == true)
{
    return false;
}
```

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+
+            if (typeName.Parent.IsKind(SyntaxKind.VariableDeclaration) &&
+                typeName.Parent.Parent.IsKind(SyntaxKind.LocalDeclarationStatement, SyntaxKind.ForStatement, SyntaxKind.UsingStatement))
+            {
```
Might be cleaner to use `IsParentKind(...) here.

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+
+                // implicitly typed variables cannot be constants.
+                var localDeclarationStatement = variableDeclaration.Parent as LocalDeclarationStatementSyntax;
+                if (localDeclarationStatement != null && localDeclarationStatement.IsConst)
```
```C#
if (localDeclarationStatement?.IsConst == true)
{
    return false;
}
```

You could even inline the `as LocalDeclarationStatementSyntax` if you want and get rid of the local altogether.

if ((variableDeclaration.Parent as LocalDeclarationStatementSyntax)?.IsConst == true)
{
    return false;
}
```

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+                    return false;
+                }
+
+                var variable = variableDeclaration.Variables.Single();
```
Does this explode on `int x = 0, y = 1;'?

balajikris replied:
(ShouldAnalyzeVariableDeclaration)[https://github.com/dotnet/roslyn/pull/8045/files#diff-38d5dc0c4ed785f53fde6c664f86c56dR252] filters out such declarations, as you could not use var on them.

DustinCampbell replied:
Cool -- thanks!

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        protected override bool AnalyzeVariableDeclaration(TypeSyntax typeName, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken, out TextSpan issueSpan)
+        {
```
`TryAnalyzeVariableDeclaration(...)`

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+            issueSpan = default(TextSpan);
+
+            // If it is already var, return.
+            if (typeName.IsTypeInferred(semanticModel))
```
I just took a look at the [`IsTypeInferred(...)`](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Extensions/TypeSyntaxExtensions.cs,88192edfd2833ed8) function and I'm not sure it's correct. It looks to me like it'll return true for a type, `var<T>`. It's esoteric, but could you check? If it does fail on that, please add a test and fix the `IsTypeInferred(...)` function.

balajikris replied:
forgot to update.. it works without any bugs, I've added a test for the said case.

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+                    issueSpan = candidateIssueSpan;
+                }
+            }
+            else if (typeName.Parent.IsKind(SyntaxKind.ForEachStatement))
```
`IsParentKind(...)`

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+        /// false, if implicit typing cannot be used.
+        /// true, otherwise.
+        /// </returns>
+        protected override bool AnalyzeAssignment(SyntaxToken identifier, TypeSyntax typeName, EqualsValueClauseSyntax initializer, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
```
Probably needs a better name that better implies the semantics of the result. E.g. `CanAssignmentBeImplicitlyTyped(...)`

balajikris replied:
Nice, thank you!

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,148 @@
+
+            // variables declared using var cannot be used further in the same initialization expression.
+            if (initializer.DescendantNodesAndSelf()
+                    .Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText.Equals(identifier.ValueText))
```
```C#
.Where(n => (n as IdentifierSyntax)?.Identifier.ValueText.Equals(identifier.ValueText) == true)
```

paul1956 commented:
I opened bug #7698 with specific examples. I also posted a sample VSIX with source under #7750 that has multiple workarounds for the various places I have found that you get the wrong result. I have tested it against Roslyn Update 1 source and found 100's of places the code finds wrong results out of 56k possible fixes.  Because of the OOM issue I have only tried to change a few 1,000 locations. What is strange is the debugger displays the correct result in the summary but has exceptions when trying to show individual fields.

paul1956 commented:
@DustinCampbell the code that identifies and fixes some of the errors is below.
```
    Private Function FixExpressionType(expressionType As ITypeSymbol) As TypeSyntax
        Dim ExpressionTypeString As String = expressionType.ToDisplayString
        Dim Index As Integer = ExpressionTypeString.IndexOf(" As ")
        If Index > 0 Then
            ExpressionTypeString = ExpressionTypeString.Substring(0, Index).Trim & ")"
        End If
        Return ParseTypeName($" {ExpressionTypeString}")
    End Function
```
I need to special case two returns I don't know how to deal with because Visual Studio errors on that is produced.
```
If variableTypeInfo.ConvertedType.ToString.Contains({"anonymous type:", "<generated method>"})
```
I also need to spcial case the following SyntaxKinds and I currently only know how to handle the first 2 well
```
SyntaxKind.AwaitExpression
SyntaxKind.CollectionInitializer

SyntaxKind.NothingLiteralExpression

SyntaxKind.FunctionLambdaHeader,
SyntaxKind.MultiLineFunctionLambdaExpression,
SyntaxKind.MultiLineSubLambdaExpression,
SyntaxKind.SingleLineFunctionLambdaExpression,
SyntaxKind.SingleLineSubLambdaExpression,
SyntaxKind.SubLambdaHeader
```

rchande commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,97 @@
+                return;
+            }
+
+            var codeAction = new MyCodeAction(
```
`MyCodeAction`?

balajikris replied:
This seems to be the general style among all our analyzer/codeaction code. Just copied the same style.

DustinCampbell replied:
I suspect the existing style was also created through copy-and-paste. :smile:

rchande commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
```
Can this type be deleted?

balajikris replied:
same reason as above.

rchande commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseExplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,91 @@
+            return true;
+        }
+
+        protected override bool AnalyzeAssignment(SyntaxToken identifier, TypeSyntax typeName, EqualsValueClauseSyntax initializer, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
```
`AssignmentIsAnonymousOrErrorType`?

rchande commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseExplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,91 @@
+            // cases :
+            //        var anon = new { Num = 1 };
+            //        var enumerableOfAnons = from prod in products select new { prod.Color, prod.Price };
+            var declaredType = semanticModel.GetTypeInfo(typeName, cancellationToken).Type;
```
Does calling `GetTypeInfo` on the type name versus the initializer expression return different information? If not, you could probably call it on just one and check that for being an error/containing anonymous types.

balajikris replied:
I've tried it and it isn't sufficient.

DustinCampbell commented on src/Features/CSharp/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1 +0,0 @@
-
```
Huh. weird diff

balajikris replied:
that's me. made a change and backed out, but some whitespace differences stuck. I likely need to scrub some commits and rewrite history.

DustinCampbell commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,71 @@
+            context.RegisterCodeFix(codeAction, context.Diagnostics.First());
+        }
+
+        private static Task<Document> ReplaceTypeWithVar(CodeFixContext context, Document document, SyntaxNode root, SyntaxNode node)
```
`context` is not used.

balajikris commented:
Since the last round of feedback, I've addressed all comments/concerns and have added enforcement styles (none, info, warning, error).

I'd appreciate another round of feedback to help me get this in with currently implemented functionalities.. I'll incrementally address further functionalities and feedback.

/cc: @DustinCampbell @rchande @Pilchie

balajikris commented on src/VisualStudio/Core/Impl/Options/OptionPreviewControl.xaml:
```diff
@@ -35,6 +38,55 @@
+                                        Margin="20, 0, 0, 0"
+                                        Grid.Column="0"/>
+
+                        <!-- TODO: remove hardcoded width -->
```
Yep, I'll fix this before checking in. Pls ignore.

balajikris commented:
I have an alternate prototype for the UI than what I already submitted here. The current checkbox style might be subtle in the sense that if unchecked and notification style set to anything other than None, it would enforce the styling by suggesting the `use explicit type` code fix, whereas this UI makes it mighty clear. thoughts?

![image](https://cloud.githubusercontent.com/assets/4213867/13201698/39fe77a0-d831-11e5-964f-d34e6ff3fb56.png)

balajikris commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseExplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,105 @@
+        {
+        }
+
+        protected override bool IsStylePreferred(SyntaxNode declarationStatement, SemanticModel semanticModel, OptionSet optionSet, State state, CancellationToken cancellationToken)
```
`declarationStatement` seems to be unused. will remove.

balajikris commented:
PR temporarily nuked due to bad merge and push.. resuscitating now...

balajikris commented:
and we're back...

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Traits.cs:
```diff
@@ -73,6 +73,8 @@ public static class Features
             public const string CodeActionsSpellcheck = "CodeActions.Spellcheck";
             public const string CodeActionsSuppression = "CodeActions.Suppression";
             public const string CodeActionsUseAutoProperty = "CodeActions.UseAutoProperty";
+            public const string CodeActionsUseImplicitTyping = "CodeActions.UseImplicitTyping";
```
We have two separate code actions?  That seems odd to me.  Why not one code action that simply applies the appropriate change?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeaturesResources.resx:
```diff
@@ -437,4 +437,16 @@
     <value>Delegate invocation can be simplified.</value>
   </data>
+  <data name="UseExplicitTyping" xml:space="preserve">
+    <value>Use explicit typing instead of 'var'</value>
```
i would far prefer you make the key:  "Use_explicit_typeing_instead_of_var".  Now i can easily find the key if i know the loc'ed string.

Also, "typing" sounds very weird.  It sounds like you're inflecting the verb "to type".  I'd really like if it just said "Use explicit type instead of 'var'".

DustinCampbell replied:
I'd prefer we make a blanket change across the whole code base rather than have a one-off on naming convention for this particular feature. Maybe file an issue to come to agreement on the style and make the change everywhere?

CyrusNajmabadi replied:
Sure.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeaturesResources.resx:
```diff
@@ -437,4 +437,16 @@
+    <value>Use explicit typing</value>
+  </data>
+  <data name="UseImplicitTyping" xml:space="preserve">
+    <value>use 'var' instead of typename</value>
```
"instead of explicit type".

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeaturesResources.resx:
```diff
@@ -437,4 +437,16 @@
+    <value>use 'var' instead of typename</value>
+  </data>
+  <data name="UseImplicitTypingDiagnosticTitle" xml:space="preserve">
+    <value>Use implicit typing</value>
```
"Use implicit type" or "Use 'var' instead of explicit type"

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,96 @@
+                             SyntaxKind.IdentifierName,
+                             SyntaxKind.GenericName,
+                             SyntaxKind.AliasQualifiedName) == false)
+            {
```
Isn't this just "is TypeSyntax"?

CyrusNajmabadi replied:
You also don't handle pointer type or just a plain old qualified type like A.B

CyrusNajmabadi replied:
Wait, if this is "use explicit type", wouldn't this always be called on a 'var' node.  Why are you checking all these other kinds?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,96 @@
+            else
+            {
+                Debug.Assert(false, $"unhandled kind {declarationContext.Kind().ToString()}");
+                return null;
```
Returning null will likely be very bad.  Either throw, or return the actualy document you started with.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,71 @@
+                return;
+            }
+
+            var node = token.GetAncestors<SyntaxNode>().First(n => n.Span.Contains(span));
```
Just call root.FindNode.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,71 @@
+                  node.IsKind(SyntaxKind.ArrayType) ||
+                  node.IsKind(SyntaxKind.IdentifierName) ||
+                  node.IsKind(SyntaxKind.GenericName) || 
+                  node.IsKind(SyntaxKind.AliasQualifiedName)))
```
look for "is TypeSyntax".

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,71 @@
+            }
+
+            var codeAction = new MyCodeAction(
+                                CSharpFeaturesResources.UseImplicitTyping,
```
indentatoin busted.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,71 @@
+            context.RegisterCodeFix(codeAction, context.Diagnostics.First());
+        }
+
+        private static Task<Document> ReplaceTypeWithVar(Document document, SyntaxNode root, SyntaxNode node)
```
Needs to end with Async.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,71 @@
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
```
All this code can be moved into the async fix function you have.  That keeps your registration code as lightweight as possible

CyrusNajmabadi replied:
This comment still applies.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,71 @@
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(span.Start);
+            if (!token.Span.IntersectsWith(span))
```
You shouldn't have to check for these sorts of things.  You're hte only one producing these diagnostic IDs, so you know the tree and span and whatnot are good.  We only need to check for IDs we get from the compiler as we have no idea what guarantees they provide about them.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeStyle/CSharpCodeStyleOptions.cs:
```diff
@@ -13,6 +14,11 @@ internal static class CSharpCodeStyleOptions
         public static readonly Option<bool> UseVarWhenDeclaringLocals = new Option<bool>(FeatureName, "UseVarWhenDeclaringLocals", defaultValue: true);
+
+        public static readonly Option<SimpleCodeStyleOption> UseVarForIntrinsicTypes = new Option<SimpleCodeStyleOption>(FeatureName, "UseImplicitTypingForIntrinsics", defaultValue: SimpleCodeStyleOption.Default);
+        public static readonly Option<SimpleCodeStyleOption> UseVarWhenTypeIsApparent = new Option<SimpleCodeStyleOption>(FeatureName, "UseImplicitTypingWhereApparent", defaultValue: SimpleCodeStyleOption.Default);
```
these names don't match:

UseImplicitTypingWhereApparent
UseVarWhenTypeIsApparent 

Pick something and use it consistently everywhere.

Also consider using nameof here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+        {
+            private readonly Dictionary<TypingStyles, DiagnosticSeverity> _styleToSeverityMap;
+
+            public TypingStyles StylePreferences { get; private set; }
```
Is TypingStyles an enum?  if so, we don't generally use the plural form with this.  this should probably just be TypeStyle

CyrusNajmabadi replied:
property name should probably match type name.  so change type ot StylePreference or change property to TypingStyle

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+            {
+                this.StylePreferences = GetCurrentTypingStylePreferences(optionSet);
+
+                IsTypingApparentInContext =
```
this can just be done with an &&

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+            /// </summary>
+            private bool IsTypeApparentInDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, TypingStyles stylePreferences, CancellationToken cancellationToken)
+            {
+                var initializer = variableDeclaration.Variables.Single().Initializer;
```
how do yo uknow there's only one variable?

balajikris replied:
We cant use `var` on a declaration that has more than one variable, so they're filtered out before it reaches here. I can add an assert if you like.

DustinCampbell replied:
An assert isn't necessary. `Single(...)` will throw if there's ever more than one.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+
+            private bool IsIntrinsicType(SyntaxNode declarationStatement) =>
+                declarationStatement.IsKind(SyntaxKind.VariableDeclaration)
+                ? ((VariableDeclarationSyntax)declarationStatement).Variables.Single().Initializer.Value.IsAnyLiteralExpression()
```
how do you know you only have a single variable?

this could just be an &&

balajikris replied:
same argument as above. I expect declarations with more than 1 variable to be filtered out before this happens.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+                ? ((VariableDeclarationSyntax)declarationStatement).Variables.Single().Initializer.Value.IsAnyLiteralExpression()
+                : false;
+
+            private ExpressionSyntax GetInitializerExpression(EqualsValueClauseSyntax initializer) =>
```
What about 'unchecked'?  What about parentheses?

balajikris replied:
This is the same type used by the compiler for both Checked and Unchecked expressions, the keyword alone is different. Parenthesis don't matter here, as I'm only trying to infer the type of the expression

CyrusNajmabadi replied:
i don't understand.  Why does this need to be unwrapped then?  If you're only trying to figure out the type, that should still work even if you call it on a checked/unchecked expressoin right?  Those nodes have the same type as their underlying expression, don't they?

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+
+            private bool IsPossibleCreationMethod(IMethodSymbol methodSymbol, ITypeSymbol declaredType, ITypeSymbol typeInInvocation)
+            {
+                var isTypeInfoSame = false;
```
What about just a static method on the type that returns something of the same type?

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+                // Pattern: Method name is an exact match of one of these well known creation methods
+                //          and method is a member of type being created.
+                // cases: node.With, Tuple.Create, State.Generate
+                if (methodSymbol.Name.Equals("With", StringComparison.Ordinal)
```
Not sure how i feel about any of this.

davkean replied:
I agree, we should come up with an algorithm for determining these - not hardcode a set of names.

balajikris replied:
you wouldn't like them to be simplified? or you are not happy with the checks performed? I'm guessing its the former.

CyrusNajmabadi replied:
I don't like the hardcoding of the names.  Though 'with' may be acceptable since we're stating that we're going to codify that pattern at the language level.

I think a more suitable approach might be:

1. If we have a static method invoked off a type, and that static method returns an instance of the same type, then the type is apparent.

i.e.  "int.Parse" is apparent, not because it's called "Parse" but because it is on "int" and returns "int".

balajikris replied:
That is infact what I'm doing. I'm checking if its a method symbol, if it returns the same type in which it is actually declared in.
I even started with the static method check, then relaxed to incorporate other patterns like `TextSpan.FromBounds` or `Tuple.Create` which didn't fit naturally.

If i lose the string checks, then it does what you ask for.

CyrusNajmabadi replied:
Sorta.

I'm explicitly stating that i want this just for static methods (since the type is apparent in the the type name).

"With" is somewhat special since it'll be an instance method.  But i'm somewhat ok with it since we're codifying "Withers" as a pattern that we use to produce new instances of the same type.

Note those: i'm not sure if "With" actually fits the definition of an apperant type.

if i have:

```c#
var x = blooby.With(...);
```

The type of x is not really apperant at all.  As opposed to:

```c#
var x = Snarfle.SomethingThatReturnsASnarfle();
```

For a type to be apperant, i would expect that i could actually *see* it on the right side somewhere.  in the "With" case, i don't see how that's so.

balajikris replied:
>  "int.Parse" is apparent, not because it's called "Parse" but because it is on "int" and returns "int"
Not really. The fact that it returns an `int` won't be apparent after we replace the declaration with `var`. It is apparent because of the verb `Parse` or `With`.

CyrusNajmabadi replied:
yes it would be.  Because i can see "int" right there on the right side.  i.e.:

```
var v = int.Parse
```

it's not parse that tells me i'm getting an int.  int's that it's "int" that i'm calling this on.

I think we can run a simple test with the .net framework.  Just enumerate the types that have static methods that return the same type as the container.  See if these are generally factory methods or if there are sufficient cases where they're not.

My bias tells me what result we'll get.  But i'm happy to let us actually try.

With this approach then we see appropriate apparant typing for things like ```XElement.Load``` and other cases.

CyrusNajmabadi replied:
Ok.  I just ran a test on all the .net refernece assemblies that start with System.  Here are cases it found that are good and that make sense:

```
WorkflowSymbol.Decode
BindingRestrictions.GetTypeRestriction
MemoryMappedFile.OpenExisting
CngKey.Import
SqlMetaData.InferFromValue
DbProviderServices.GetProviderServices
Principal.FindByIdentity
Domain.GetCurrentDomain
Process.GetProcessById
ContractDescription.GetContract
System.Xml.Linq.XElement.Load
etc. etc.
```

Plus there are *tons* of hits that start with "Create".

Overall, this seems like a fairly reasonable heuristic.

CyrusNajmabadi replied:
Also note: you may end up wanting to match type names instead of using type equality.  that will ensure that if the user calls "Tuple.Create(0, true)" that we'll consider that sufficient.  This is because Tuple.Create returns a ```Tuple<X,Y>``` which is not the same type as ```Tuple```.

balajikris replied:
Thanks for the heuristic and the facts. I'll implement this.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,237 @@
+                // case: Convert.ToString or iConvertible.ToChar
+                if (methodSymbol.Name.Equals("To" + declaredTypeName, StringComparison.Ordinal))
+                {
+                    var convertType = semanticModel.Compilation.ConvertType();
```
Why bother with the check for Convert or IConvertible.  How about if you ever have something like "Point ToPoint", then you say "it's clear from the name..."

balajikris replied:
I'll handle the `Point.ToPoint` case with the above heuristic we discussed about static methods. This however _is _ required to handle code where we literally use `Convert.ToString` or an instance of iConvertible and a To method(). Notice how it doesn't fall under the general heuristic we talked about above.

CyrusNajmabadi replied:
Sorry, i wasn't clear. I'm not talking about ```Point.ToPoint```.  I'm talkign about   ```Point AnyTypeAtAll.ToPoint```

My point (no pun intended) is that you don't need to special case IConvertible or Convert.  Just say:

1.  If we have a method called ```ToXXX``` and that method returns a type called ```XXX```, then that means the type name is apparent.

That's it.  Now the code works all over the place and not just for these two types.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,163 @@
+            None = 0,
+            VarForIntrinsic = 1 << 0,
+            VarWhereApparent = 1 << 1,
+            VarWherePossible = 1 << 2,
```
More cases where these names are ever so slightly off from teh exact same concept elsehwere.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,163 @@
+            _diagnosticId = diagnosticId;
+            _title = title;
+            _message = message;
+            _noneDiagnosticDescriptor = new Lazy<DiagnosticDescriptor>(() => CreateDiagnosticDescriptor(DiagnosticSeverity.Hidden));
```
This lazyness is not saving us anything.  We now have more allocatoins because of hte lazy and hte delegate.  just new up the diagnostic descriptor.  Having 4 extra objects for them is not a problem.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,163 @@
+            }
+            else
+            {
+                Debug.Assert(false, $"called in for unregistered node kind {declarationStatement.Kind().ToString()}");
```
Just throw.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,163 @@
+
+                if (shouldAnalyze)
+                {
+                    state = State.Generate(declarationStatement, context.SemanticModel, optionSet, isVariableDeclarationContext: true, cancellationToken: context.CancellationToken);
```
introduce local for things like hte semantic model and cancellation token.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,163 @@
+        private bool ShouldAnalyzeVariableDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // var is applicable only for local variables.
+            if (variableDeclaration.Parent.IsKind(SyntaxKind.FieldDeclaration) ||
```
Don't do negative checks.  Use a positive check.  i.e. "variableDeclaration.parent.isKind(LocalDeclarationStatement)"

davkean replied:
Why? These are basically earlier exits, and are subjective. :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,154 @@
+            }
+
+            var candidateReplacementNode = SyntaxFactory.IdentifierName("var")
+                                                .WithLeadingTrivia(typeName.GetLeadingTrivia())
```
i don't think you need to bother with the trivia.

DustinCampbell replied:
Sure he does.

```C#
// My cool variable
int var = int.Parse("42");
```

CyrusNajmabadi replied:
But this is only for speculative purposes.  We're literally only asking "does 'var' bind to anything here".  That can be asked without caring about trivia.

Note: it's unclear if we even need to lookup using a node.  The compiler already has a "LookupSymbols" api that you could just pass "var" to to see if it binds to anything.

CyrusNajmabadi replied:
I'm lost. I'm still not seeing where this node is used *beyond* just seeing how it is bound speculatively.  As such, i don't see why it needs trivia.

balajikris replied:
you're right. I'm returning the span, the node isn't used for construction -- clearly, thats handled by the fixer and this is in the analyzer.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,154 @@
+                return false;
+            }
+
+            if (typeName.Parent.IsKind(SyntaxKind.VariableDeclaration) &&
```
repeated conditoin, extract a helper.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,154 @@
+
+        protected override bool TryAnalyzeVariableDeclaration(TypeSyntax typeName, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken, out TextSpan issueSpan)
+        {
+            issueSpan = default(TextSpan);
```
i don't like this.  it masks cases where you might miss a flow path.  Don't set here, but set later on when appropriate.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,154 @@
+            if (declaredType != null
+                && (declaredType.TypeKind == TypeKind.Delegate
+                || declaredType.TypeKind == TypeKind.Dynamic))
+            {
```
this would be a lot clearer (to me) as:

```
if (declaredType != null &&
    (... || ...))
{
}

Right now hte alignment makes it seem as if the && and || are part o the same expressoin.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,154 @@
+            var expression = initializer.Value;
+
+            // var cannot be assigned null
+            if (expression.IsKind(SyntaxKind.NullLiteralExpression))
```
have you appropriately unwrapped the expression of parentheses first?

balajikris replied:
Nice catch, will fix.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/PredefinedCodeFixProviderNames.cs:
```diff
@@ -36,5 +36,7 @@ internal static class PredefinedCodeFixProviderNames
         public const string AddOverloads = "Add Overloads to member";
         public const string AddNew = "Add new keyword to member";
+        public const string UseImplicitTyping = "Use var";
+        public const string UseExplicitTyping = "Use typename instead of var";
```
use use nameof here.

DustinCampbell replied:
Huh? But the strings are different.

CyrusNajmabadi replied:
Yes.  That's sort of the problem.  We shoudl not have these be inconsistent.  Either:

1. The const name should match the text.  i.e. "const string UseTypeNameInsteadOfVar" or
2. The value should match the name.  i.e. "const string UseExplicitTyping = nameof(UseExplicitTyping)"

CyrusNajmabadi commented on src/Features/Core/Portable/CodeStyle/NotificationOption.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    internal class NotificationOption
```
Why is this type needed?  It sems to just wrap DiagnosticSeverity...

balajikris replied:
for WPF binding.. If i use the enum, i need an ObjectDataProvider to retrieve values from `Enum.GetValues` method. whereas by making a type and having properties and ToString() on the type, I'm making it easier for data binding.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs:
```diff
@@ -520,5 +561,52 @@ private void SetBooleanOption(PerLanguageOption<bool> key, int value)
+            var optionSet = _optionService.GetOptions();
+            switch (value)
+            {
+                case 0:
```
Where did these numbers come from?  Why can't we encode this in a much clearer manner?

balajikris replied:
encoded right above in `GetUseVarOption`. basically they range from 0 to 2*(DiagnosticSeverity.Length) [4 for use var, 4 for use explicit type]

CyrusNajmabadi replied:
Oh yeah, i passionately hate this.  Please encode differently.  it could be a basic string with the two enums encoded in it.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/CSharpSettingsManagerOptionSerializer.cs:
```diff
@@ -274,7 +296,67 @@ public override bool TryPersist(OptionKey optionKey, object value)
+        {
+            switch (useVarOptionValue)
+            {
+                case 0:
```
Oh man... do these two switches need to be in sync?  If so, please let's not do this.  Let's just have a string representation, and we can have SImpleCodeStyleOption serialize/deserialize itself from that.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs:
```diff
@@ -2291,6 +2291,18 @@ public static SyntaxKind GetPredefinedKeywordKind(SpecialType specialType)
         {
+            var awaitExpression = node as AwaitExpressionSyntax;
+            if (awaitExpression != null && awaitExpression.Expression != null)
+            {
```
We're going far out of ht realm of what this function was for.  this was for:  I have a dotted name.  I want the rightmost name.

CyrusNajmabadi replied:
I'm going to still push back on this.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs:
```diff
@@ -2291,6 +2291,18 @@ public static SyntaxKind GetPredefinedKeywordKind(SpecialType specialType)
+            }
+
+            var invocationExpression = node as InvocationExpressionSyntax;
+            if (invocationExpression != null && invocationExpression.Expression != null)
```
This is especially concerning.  An invocation doesn't even have a rightmost name.  Because it's got more stuff to the rigth.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Shared/Extensions/ICompilationExtensions.cs:
```diff
@@ -133,5 +133,15 @@ public static INamedTypeSymbol SuppressMessageAttributeType(this Compilation com
+
+        public static INamedTypeSymbol IConvertibleType(this Compilation compilation)
+        {
+            return compilation.GetTypeByMetadataName("System.IConvertible");
```
Does your code that uses this ensure that it handles null properly?

balajikris replied:
yep. thanks.

CyrusNajmabadi replied:
These aren't needed anymore, right?

DustinCampbell commented on src/Features/CSharp/Portable/CodeStyle/CSharpCodeStyleOptions.cs:
```diff
@@ -13,6 +14,11 @@ internal static class CSharpCodeStyleOptions
     {
         internal const string FeatureName = "CSharpCodeStyle";
 
+        // TODO: get sign off on public api changes.
```
Signed off. :smile:

DustinCampbell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,163 @@
+            // implicitly typed variables cannot have multiple declarators and
+            // must have an initializer.
+            if (variableDeclaration.Variables.Count > 1 ||
+                !variableDeclaration.Variables.Single().Initializer.IsKind(SyntaxKind.EqualsValueClause))
```
Note that this will throw if there are 0 variables due to the call to `Single(...)`. It doesn't look like that was your intent.

DustinCampbell commented:
Assuming pending feedback is address :+1: -- we can work on further refinement once the initial feature is checked in.

balajikris commented:
retest vsi please

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitTyping/UseExplicitTypingTests.cs:
```diff
@@ -0,0 +1,906 @@
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Diagnostics.UseExplicitTyping
```
"Typing" :(

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/UseImplicitTyping/UseImplicitTypingTests.cs:
```diff
@@ -0,0 +1,1381 @@
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Diagnostics.UseImplicitTyping
```
"Typing" -> "Type"/

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Traits.cs:
```diff
@@ -73,6 +73,8 @@ public static class Features
             public const string CodeActionsSuppression = "CodeActions.Suppression";
             public const string CodeActionsUseAutoProperty = "CodeActions.UseAutoProperty";
+            public const string CodeActionsUseImplicitTyping = "CodeActions.UseImplicitTyping";
+            public const string CodeActionsUseExplicitTyping = "CodeActions.UseExplicitTyping";
```
"Typing" -> "Type".

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeatures.csproj:
```diff
@@ -86,6 +86,8 @@
     <Compile Include="CodeFixes\SimplifyTypeNames\SimplifyTypeNamesCodeFixProvider.SimplifyTypeNamesFixAllProvider.cs" />
     <Compile Include="CodeFixes\SimplifyTypeNames\SimplifyTypeNamesCodeFixProvider.cs" />
     <Compile Include="CodeFixes\Suppression\CSharpSuppressionCodeFixProvider.cs" />
+    <Compile Include="CodeFixes\UseImplicitTyping\UseExplicitTypingCodeFixProvider.cs" />
```
"Typing" -> "Type".

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,88 @@
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.UseImplicitTyping
```
"Typing" -> "Type"

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,88 @@
+            var span = context.Span;
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(span.Start);
```
You can just call FindNode.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,88 @@
+            }
+            else
+            {
+                Debug.Assert(false, $"unhandled kind {declarationContext.Kind().ToString()}");
```
I'd prefer Contract.Fail

CyrusNajmabadi replied:
or throw new InvalidOperationException.  Nobody runs debug.  so these asserts go nowhere.

balajikris replied:
No, I dont want VS to crash in the case where i didn't handle a declaration node. Its alright if we dont show the fix there, someone will file a bug, as will they if it crashes. But somehow i don't like the Idea of taking out VS for this purpose and hey, after the bug is identified, the crash becomes really painful until we push a fix. This isn't a there is nothing we can do after this, so crash kind of scenario.

CyrusNajmabadi replied:
It won't take down VS.  We''ll catch the exception, report it and disable the fixer.  We'll also let the user know so they can reenable if they want.

Throwing is the right thing to do here.  Otherwise, there's almost no way we'll find out about issues here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,88 @@
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument) :
+                base(title, createChangedDocument)
```
You need to pass 'title' along for the equivalence key.

Also, did you implement the batch fix provider?  we'll want that for this feature.

balajikris replied:
I forgot about fix all :cry: -- will do.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,66 @@
+                return;
+            }
+
+            var node = root.FindNode(span, getInnermostNodeForTie: true);
```
Can still just call FindNode.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseImplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,66 @@
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument) :
+                base(title, createChangedDocument)
```
need to pass the title along as the equivalent key.  also have you implemented the batch fill all stuff?

CyrusNajmabadi replied:
you'll also want tests for fix all.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -0,0 +1,171 @@
+        private bool ShouldAnalyzeVariableDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // var is applicable only for local variables.
+            if (variableDeclaration.IsParentKind(SyntaxKind.FieldDeclaration,
```
the check if it's parented by a local variable statement or not :)

balajikris replied:
Lol. its either a whitelist or blacklist, i guess i did a blacklist of things that are not suited.

CyrusNajmabadi replied:
the problem with a blacklist is that you might forget something (and then who knows what happens).  With a whitelist you specify precisely what you handle.  Please whitelist here.

balajikris replied:
While I agree with that statement, in this case here, my blacklist is precise and simple - don't act on fields or declarations with more than 1 variable or the ones without an initializer.

> the check is it's parented by a local variable declaration
and a for statement
and a using statement.

I am not strongly attached to a style here, just sticking to what seems to be precise and elegant. I've left it as such, but if you feel very strongly, I'll change ;)

CyrusNajmabadi replied:
This needs to be a whitelist.  How do you know that just filtering out fields will be enough?  What happens when the compiler adds another location where VairableDeclarations exist?

balajikris replied:
If the compiler adds another location, I or whoever supports ide features has to touch this lost -- to specifically include or exclude that kind. But yeah, with a whitelist missing a kind, it's restricted to feature not working there and we get away without a crash. I'll change this in a bit and take your if branch that yields the :+1 ;)

Sent from Outlook Mobile<https://aka.ms/blhgte>

From: CyrusNajmabadi
Sent: Wednesday, February 24, 5:06 PM
Subject: Re: [roslyn] Feature: Suggest implicit or explicit typing code styles. (#8045)
To: dotnet/roslyn
Cc: Balaji Krishnan

In src<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>/Features/<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>CSharp<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>/Portable/Diagnostics/Analyzers/<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>CSharpTypingStyleDiagnosticAnalyzerBase.cs<https://github.com/dotnet/roslyn/pull/8045#discussion_r54034687>:

> + > + if (TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken, out diagnosticSpan)) > + { > + var descriptor = _severityToDescriptorMap[state.GetDiagnosticSeverityPreference()]; > + context.ReportDiagnostic(CreateDiagnostic(descriptor, declarationStatement, diagnosticSpan)); > + } > + } > + } > + > + private Diagnostic CreateDiagnostic(DiagnosticDescriptor descriptor, SyntaxNode declaration, TextSpan diagnosticSpan) => > + Diagnostic.Create(descriptor, declaration.SyntaxTree.GetLocation(diagnosticSpan)); > + > + private bool ShouldAnalyzeVariableDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken) > + { > + // var is applicable only for local variables. > + if (variableDeclaration.IsParentKind(SyntaxKind.FieldDeclaration,

This needs to be a whitelist. How do you know that just filtering out fields will be enough? What happens when the compiler adds another location where VairableDeclarations exist?

-
Reply to this email directly or view it on <https://github.com/dotnet/roslyn/pull/8045/files#r54034687> GitHub<https://github.com/dotnet/roslyn/pull/8045/files#r54034687>.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/PredefinedCodeFixProviderNames.cs:
```diff
@@ -36,5 +36,7 @@ internal static class PredefinedCodeFixProviderNames
         public const string Suppression = nameof(Suppression);
         public const string AddOverloads = "Add Overloads to member";
         public const string AddNew = "Add new keyword to member";
+        public const string UseImplicitTyping = nameof(UseImplicitTyping);
```
change these.  these are part of our public API.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/CSharpVSResources.resx:
```diff
@@ -411,4 +411,16 @@
+    <value>Use 'var' for built-in types</value>
+  </data>
+  <data name="UseVarWhenPossible" xml:space="preserve">
+    <value>Use 'var' elsewhere</value>
```
I hate that the key and value have no relation.  At least call this UseVarElsewhere.  I want to be able to read the key in the code and know what stirng will be shown to the user.

balajikris replied:
dont worry about this - this is the UI that shows in tools/options. I'm not done here, need to do lots of changes here.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/CSharpSettingsManagerOptionSerializer.cs:
```diff
@@ -274,7 +296,42 @@ public override bool TryPersist(OptionKey optionKey, object value)
+            }
+            else if (optionKey.Option == CSharpCodeStyleOptions.UseImplicitTypeWhereApparent)
+            {
+                return PersistUseVarOption(Style_UseVarWhenTypeIsApparent, value);
```
This name doesn't match the style option.    One says UseVar and the other says UseImplicitType

balajikris replied:
yeah, my bad. I only touched user facing strings and variables. which is why the mismatch. I'll fix.

CyrusNajmabadi commented:
I :+1: *if* you change to a whitelist :)

Pilchie commented on src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
```
If you don't error out, it would be better to move this into HandleDeclarationAsync.  That will mean we don't have to get the syntax tree in order to show the lightbulb - only to expand it.

Pilchie commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,261 @@
+                if (declaredType.GetTypeArguments().Length > 0 ||
+                    typeInInvocation.GetTypeArguments().Length > 0)
+                {
+                    return declaredType.Name.Equals(returnType.Name);
```
I would rather use `GetOriginalType().Equals()` here.

Pilchie commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,261 @@
+                    var typeName = memberName.GetLeftSideOfDot();
+                    return IsPossibleCreationOrConversionMethod(methodSymbol, declaredTypeSymbol, semanticModel, typeName, cancellationToken);
+                }
+
```
What about enums (or in fact any public static members of a type)? Like:

```C#
enum Color { Red, Green, Blue }
...
var color = Color.Red;
```

Just to make it tough, what about fancy patterns like `Color c = Colors.Blue` in the WPF case, or event `Brush b = SystemBrushes.Background`?

balajikris replied:
Nice, I'll add this.

Pilchie commented:
I know I'm late to the party, but LGTM.
